In [86]:
import pandas as pd
import numpy
import matplotlib

import seaborn
import  plotly
import  scipy
import  statsmodels

#returns all the rows and columns for the dataframe

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

# Force pandas to display full numbers instead of scientific notation
# pd.options.display.float_format = '{:.0f}'.format

# Library to suppress warnings
import warnings
warnings.filterwarnings('ignore')

# Dataset Description:
The Public School Characteristics 2020-2021 dataset is provided by the National Center for Education Statistics. This dataset provides information on schools across the U.S and icnludes data on school identifiers, location details, enrollment information, grade leves, Title 1 funding status, and school types such as charter or magnent schools. This dataset also includes information surrounding student demopgraphics, and classifes schools based on their geographic location such as urban, suburban, town, or rual. 

### Link to Dataset: https://catalog.data.gov/dataset/public-school-characteristics-2020-21-6120a



 May need this dataset to further understand poverty levels if different schools:

Title: School Neighborhood Poverty Estimates, 2020-21
Link:https://catalog.data.gov/dataset/school-neighborhood-poverty-estimates-2020-21

Description: The 2020-2021 School Neighborhood Poverty Estimates are based on school locations from the 2020-2021 Common Core of Data (CCD) school file and income data from families with children ages 5 to 17 in the U.S. Census Bureau’s 2017-2021 American Community Survey (ACS) 5-year collection. The ACS is a continuous household survey that collects social, demographic, economic, and housing information from the population in the United States each month. The Census Bureau calculates the income-to-poverty ratio (IPR) based on money income reported for families relative to the poverty thresholds, which are determined based on the family size and structure. Noncash benefits (such as food stamps and housing subsidies) are excluded, as are capital gains and losses. The IPR is the percentage of family income that is above or below the federal poverty level. The IPR indicator ranges from 0 to a top-coded value of 999. A family with income at the poverty threshold has an IPR value of 100. The estimates in this file reflect the IPR for the neighborhoods around schools which may be different from the neighborhood conditions of students enrolled in schools.All information contained in this file is in the public domain. Data users are advised to review NCES program documentation and feature class metadata to understand the limitations and appropriate use of these data.

# Research Topic:


Education funding and resource allocation may contribute to disparities in student experiences across different school enviorments. This research explores how factors such as school funding, poverty levels, and racial demographics influence student-to-teacher ratios and access to educational resources in urban, suburban, and rural schools.

# Research Problem

Despite efforts to address educational inequities, disparities in teacher allocation and resource distribution remain a common trend, particularly in schools that serve predominantly Black and Hispanic students. Schools with higher poverty levels, measured by free and reduced-price lunch eligibility, may still experience larger class sizes and fewer resources compared to wealthier schools. In addition, geographic location—whether a school is in an urban, suburban, or rural area—may further contribute to these disparities. This study hypothesizes that schools with higher Black and Hispanic enrollment and greater poverty rates will have higher student-to-teacher ratios than predominantly White schools, with these disparities being most pronounced in urban areas.

# Reading My Data:

In [87]:
# Checking to make sure my file is in the right place
import os

folder_path = "/Users/remi/Documents/Masters Fall 2024/Public School Charachtertics"

# List of all files in my folder
print(os.listdir(folder_path))


['.DS_Store', 'Public School Test Scores & Demographics.csv', 'Public_School_Characteristics_2022-23.csv', 'Working files ', 'Public_School_Characteristics_2020-21.csv', '7_CCD_Glossary.pages', 'Public School charchtertics KEY.pdf', 'Public_School_Characteristics_2017-18.csv', '~$Public School Test Scores & Demographics.xlsx', 'Public School Test Scores & Demographics.xlsx']


In [88]:
file_path = "/Users/remi/Documents/Masters Fall 2024/Public School Charachtertics/Public_School_Characteristics_2020-21.csv"

df = pd.read_csv(file_path)  # Load CSV

# Viewing The Data

##The Variables:

X, Y – Likely geographic coordinates (longitude, latitude).
OBJECTID – Unique identifier for each record.
NCESSCH – National Center for Education Statistics (NCES) School ID.
SURVYEAR – Survey year (2022-23).
STABR – State abbreviation.
LEAID – Local Education Agency (LEA) ID.
ST_LEAID – State-specific LEA ID.
LEA_NAME – Name of the Local Education Agency (school district).
SCH_NAME – Name of the school.
LSTREET1, LSTREET2 – School address (street lines 1 and 2).
**LCITY – School city.
**LSTATE – School state.
LZIP, LZIP4 – School ZIP code (5-digit and extended 4-digit).
PHONE – School contact number.
**CHARTER_TEXT – Indicates whether the school is a charter school.
VIRTUAL – Indicates if the school operates as a virtual school.
GSLO, GSHI – Lowest and highest grade levels offered at the school.
SCHOOL_LEVEL – Classification of the school (elementary, middle, high).
STATUS – Operational status of the school (e.g., open, closed).
SCHOOL_TYPE_TEXT – School type (e.g., public, magnet, alternative).
SY_STATUS_TEXT – School year status.
ULOCALE – Urban locale classification (e.g., rural, suburban, urban).
NMCNTY – County name.
TOTFRL – Total students eligible for free or reduced lunch.
FRELCH – Number of students receiving free lunch.
REDLCH – Number of students receiving reduced-price lunch.
DIRECTCERT – Number of students directly certified for free meals.
PK, KG, G01-G12, G13, UG, AE – Enrollment counts by grade level.
TOTMENROL – Total male enrollment.
TOTFENROL – Total female enrollment.
TOTAL – Total school enrollment.
MEMBER – Membership count (another enrollment metric).
FTE – Full-time equivalent staff.
STUTERATIO – Student-teacher ratio.
AMALM, AMALF, AM – American Indian/Alaska Native male, female, and total counts.
ASALM, ASALF, AS – Asian male, female, and total counts.
BLALM, BLALF, BL – Black or African American male, female, and total counts.
HPALM, HPALF, HP – Native Hawaiian/Pacific Islander male, female, and total counts.
HIALM, HIALF, HI – Hispanic or Latino male, female, and total counts.
TRALM, TRALF, TR – Two or more races male, female, and total counts.
WHALM, WHALF, WH – White male, female, and total counts.
LATCOD, LONCOD – Latitude and longitude coordinates of the school.

In [113]:
df.head(7)

X        Y  OBJECTID      NCESSCH   SURVYEAR STABR   LEAID  \
0 -86.206200  34.2602         1  10000500870  2020-2021    AL  100005   
1 -86.204900  34.2622         2  10000500871  2020-2021    AL  100005   
2 -86.220100  34.2733         3  10000500879  2020-2021    AL  100005   
3 -86.221806  34.2527         4  10000500889  2020-2021    AL  100005   
4 -86.193300  34.2898         5  10000501616  2020-2021    AL  100005   
5 -86.221800  34.2533         6  10000502150  2020-2021    AL  100005   
6 -86.254100  34.5337         7  10000600193  2020-2021    AL  100006   

  ST_LEAID          LEA_NAME                           SCH_NAME  \
0   AL-101  Albertville City          Albertville Middle School   
1   AL-101  Albertville City            Albertville High School   
2   AL-101  Albertville City    Albertville Intermediate School   
3   AL-101  Albertville City      Albertville Elementary School   
4   AL-101  Albertville City  Albertville Kindergarten and PreK   
5   AL-101  Albertville City         Albertville Primary School   
6   AL-048   Marshall County       Kate Duncan Smith DAR Middle   

              LSTREET1 LSTREET2        LCITY LSTATE   LZIP LZIP4  \
0    600 E Alabama Ave      NaN  Albertville     AL  35950         
1     402 E McCord Ave      NaN  Albertville     AL  35950  2322   
2   901 W McKinney Ave      NaN  Albertville     AL  35950  1300   
3   145 West End Drive      NaN  Albertville     AL  35950         
4  257 Country Club Rd      NaN  Albertville     AL  35951  3927   
5       1100 Horton Rd      NaN  Albertville     AL  35950  2532   
6         6077 Main St      NaN        Grant     AL  35747         

           PHONE CHARTER_TEXT MAGNET_TEXT      VIRTUAL GSLO GSHI SCHOOL_LEVEL  \
0  (256)878-2341           No          No  Not Virtual   07   08       Middle   
1  (256)894-5000           No          No  Not Virtual   09   12         High   
2  (256)878-7698           No          No  Not Virtual   05   06       Middle   
3  (256)894-4822           No          No  Not Virtual   03   04   Elementary   
4  (256)878-7922           No          No  Not Virtual   PK   KG   Elementary   
5  (256)878-6611           No          No  Not Virtual   01   02   Elementary   
6  (256)728-5950           No          No  Not Virtual   05   08       Middle   

  TITLEI STITLEI  STATUS SCHOOL_TYPE_TEXT          SY_STATUS_TEXT  \
0  1-Yes   1-Yes       1   Regular school  Currently operational    
1  1-Yes   1-Yes       1   Regular school  Currently operational    
2  1-Yes   1-Yes       1   Regular school  Currently operational    
3  1-Yes   1-Yes       1   Regular school  Currently operational    
4  1-Yes   1-Yes       1   Regular school  Currently operational    
5  1-Yes   1-Yes       1   Regular school  Currently operational    
6  1-Yes   1-Yes       1   Regular school  Currently operational    

             ULOCALE           NMCNTY  TOTFRL  FRELCH  REDLCH     PK     KG  \
0   32-Town: Distant  Marshall County   332.0   332.0     0.0    NaN    NaN   
1   32-Town: Distant  Marshall County   456.0   456.0     0.0    NaN    NaN   
2   32-Town: Distant  Marshall County   330.0   330.0     0.0    NaN    NaN   
3   32-Town: Distant  Marshall County   397.0   397.0     0.0    NaN    NaN   
4   32-Town: Distant  Marshall County   181.0   181.0     0.0  113.0  443.0   
5   32-Town: Distant  Marshall County   422.0   422.0     0.0    0.0    NaN   
6  42-Rural: Distant  Marshall County   156.0   127.0    29.0    NaN    NaN   

     G01    G02    G03    G04    G05    G06    G07    G08    G09    G10  \
0    NaN    NaN    NaN    NaN    NaN    NaN  469.0  439.0    NaN    NaN   
1    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  421.0  427.0   
2    NaN    NaN    NaN    NaN  439.0  452.0    NaN    NaN    NaN    NaN   
3    NaN    NaN  435.0  473.0    NaN    NaN    NaN    NaN    NaN    NaN   
4    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5  480.0  493.0    NaN    NaN    NaN    NaN    NaN    NaN  

In [114]:
df.tail(7)

X          Y  OBJECTID       NCESSCH   SURVYEAR STABR  \
100715 -64.932456  18.352146    100716  780003000020  2020-2021    VI   
100716 -64.793916  18.330464    100717  780003000022  2020-2021    VI   
100717 -64.917602  18.341950    100718  780003000024  2020-2021    VI   
100718 -64.952483  18.338742    100719  780003000026  2020-2021    VI   
100719 -64.899024  18.354782    100720  780003000027  2020-2021    VI   
100720 -64.945940  18.336658    100721  780003000033  2020-2021    VI   
100721 -64.890311  18.318230    100722  780003000034  2020-2021    VI   

          LEAID ST_LEAID                                   LEA_NAME  \
100715  7800030   VI-001  Saint Thomas - Saint John School District   
100716  7800030   VI-001  Saint Thomas - Saint John School District   
100717  7800030   VI-001  Saint Thomas - Saint John School District   
100718  7800030   VI-001  Saint Thomas - Saint John School District   
100719  7800030   VI-001  Saint Thomas - Saint John School District   
100720  7800030   VI-001  Saint Thomas - Saint John School District   
100721  7800030   VI-001  Saint Thomas - Saint John School District   

                                SCH_NAME                     LSTREET1  \
100715  JOSEPH SIBILLY ELEMENTARY SCHOOL  14  15  16 ESTATE ELIZABETH   
100716                  JULIUS E SPRAUVE         14 18 ESTATE ENIGHED   
100717        LOCKHART ELEMENTARY SCHOOL             41 ESTATE THOMAS   
100718   ULLA F MULLER ELEMENTARY SCHOOL            7B ESTATE CONTANT   
100719   YVONNE BOWSKY ELEMENTARY SCHOOL    15B and 16 ESTATE MANDAHL   
100720        CANCRYN JUNIOR HIGH SCHOOL                  1 CROWN BAY   
100721      BERTHA BOSCHULTE JUNIOR HIGH           9 1 and 12A BOVONI   

       LSTREET2         LCITY LSTATE  LZIP LZIP4          PHONE CHARTER_TEXT  \
100715      NaN  Saint Thomas     VI   802        (340)774-7001            N   
100716      NaN    Saint John     VI   831        (340)776-6336            N   
100717      NaN  Saint Thomas     VI   802        (340)775-0820            N   
100718      NaN  Saint Thomas     VI   802        (340)774-0059            N   
100719      NaN  Saint Thomas     VI   802        (340)775-3220            N   
100720      NaN  Saint Thomas     VI   804        (340)774-4540            N   
100721      NaN  Saint Thomas     VI   802        (340)775-4222            N   

       MAGNET_TEXT       VIRTUAL GSLO GSHI SCHOOL_LEVEL TITLEI STITLEI  \
100715          No  Full Virtual   KG   06   Elementary  1-Yes   1-Yes   
100716          No  Full Virtual   PK   08   Elementary  1-Yes   1-Yes   
100717          No  Full Virtual   KG   03   Elementary  1-Yes   1-Yes   
100718          No  Full Virtual   KG   06   Elementary  1-Yes   1-Yes   
100719          No  Full Virtual   PK   05   Elementary  1-Yes   1-Yes   
100720          No  Full Virtual   04   08       Middle  1-Yes   1-Yes   
100721          No  Full Virtual   06   08       Middle  1-Yes   1-Yes   

        STATUS SCHOOL_TYPE_TEXT          SY_STATUS_TEXT          ULOCALE  \
100715       1   Regular school  Currently operational   33-Town: Remote   
100716       1   Regular school  Currently operational   33-Town: Remote   
100717       1   Regular school  Currently operational   33-Town: Remote   
100718       1   Regular school  Currently operational   33-Town: Remote   
100719       1   Regular school  Currently operational   33-Town: Remote   
100720       1   Regular school  Currently operational   33-Town: Remote   
100721       1   Regular school  Currently operational   33-Town: Remote   

                   NMCNTY  TOTFRL  FRELCH  REDLCH    PK    KG   G01   G02  \
100715  St. Thomas Island   234.0   234.0     0.0   NaN  25.0  28.0  28.0   
100716    St. John Island   221.0   221.0     0.0   9.0  14.0  26.0  20.0   
100717  St. Thomas Island   325.0   325.0     0.0   NaN  71.0  75.0  73.0   
100718  St. Thomas Island   474.0   474.0     0.0   NaN  50.0  59.0  72.0   
100719  St. Thomas Island   394.0   394.0     0.0

In [116]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100722 entries, 0 to 100721
Data columns (total 79 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   X                 100722 non-null  float64
 1   Y                 100722 non-null  float64
 2   OBJECTID          100722 non-null  int64  
 3   NCESSCH           100722 non-null  int64  
 4   SURVYEAR          100722 non-null  object 
 5   STABR             100722 non-null  object 
 6   LEAID             100722 non-null  int64  
 7   ST_LEAID          100722 non-null  object 
 8   LEA_NAME          100722 non-null  object 
 9   SCH_NAME          100722 non-null  object 
 10  LSTREET1          100719 non-null  object 
 11  LSTREET2          578 non-null     object 
 12  LCITY             100722 non-null  object 
 13  LSTATE            100722 non-null  object 
 14  LZIP              100722 non-null  int64  
 15  LZIP4             100722 non-null  object 
 16  PHONE             10

In [91]:
df.shape

(100722, 79)

In [92]:

# This is how many datapoints there are in total
101390* 77


7807030

### Preliminary Notes on Dataset

- There is a total of 7807030 datapoints in this dataset
- There are 100722 rows and 79 columns in this dataset 

In [93]:
# this is showing me the data type for all of my columns
df.info(show_counts=True, verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100722 entries, 0 to 100721
Data columns (total 79 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   X                 100722 non-null  float64
 1   Y                 100722 non-null  float64
 2   OBJECTID          100722 non-null  int64  
 3   NCESSCH           100722 non-null  int64  
 4   SURVYEAR          100722 non-null  object 
 5   STABR             100722 non-null  object 
 6   LEAID             100722 non-null  int64  
 7   ST_LEAID          100722 non-null  object 
 8   LEA_NAME          100722 non-null  object 
 9   SCH_NAME          100722 non-null  object 
 10  LSTREET1          100719 non-null  object 
 11  LSTREET2          578 non-null     object 
 12  LCITY             100722 non-null  object 
 13  LSTATE            100722 non-null  object 
 14  LZIP              100722 non-null  int64  
 15  LZIP4             100722 non-null  object 
 16  PHONE             10

### Null values 

In [94]:
#Showing me how many values are missing in each column--the .sum() just gives me the count 
df.isnull().sum()

X                        0
Y                        0
OBJECTID                 0
NCESSCH                  0
SURVYEAR                 0
STABR                    0
LEAID                    0
ST_LEAID                 0
LEA_NAME                 0
SCH_NAME                 0
LSTREET1                 3
LSTREET2            100144
LCITY                    0
LSTATE                   0
LZIP                     0
LZIP4                    0
PHONE                    0
CHARTER_TEXT             0
MAGNET_TEXT              0
VIRTUAL                  0
GSLO                     0
GSHI                     0
SCHOOL_LEVEL             0
TITLEI                   0
STITLEI                  0
STATUS                   0
SCHOOL_TYPE_TEXT         0
SY_STATUS_TEXT           0
ULOCALE                  0
NMCNTY                   0
TOTFRL               23758
FRELCH               26605
REDLCH               26605
PK                   69021
KG                   47185
G01                  46781
G02                  46742
G

In [95]:
print(df['SCHOOL_TYPE_TEXT'].value_counts())

SCHOOL_TYPE_TEXT
Regular school                  91595
Alternative Education School     5588
Special education school         1946
Career and Technical School      1593
Name: count, dtype: int64


In [96]:
print(df['SY_STATUS_TEXT'].value_counts())

# I need to get rid of 

SY_STATUS_TEXT
Currently operational                                98717
New school                                             902
School temporarily closed                              521
School to be operational within two years              438
School was operational but not reported last year       73
School has changed agency                               40
School has reopened                                     31
Name: count, dtype: int64


In [97]:
def missing(dataframe):
    print('Percentage of missing values in the dataset:\n',
          round((dataframe.isnull().sum()*100/len(dataframe)),2).sort_values(ascending=False))

In [98]:
missing(df)

Percentage of missing values in the dataset:
 TOTFENROL           100.00
TOTMENROL           100.00
G13                  99.86
AE                   99.82
LSTREET2             99.43
UG                   91.91
G12                  73.46
G11                  73.37
G10                  73.35
G09                  73.21
PK                   68.53
G07                  68.08
G08                  67.85
G06                  63.05
G05                  47.95
KG                   46.85
G04                  46.66
G03                  46.45
G01                  46.45
G02                  46.41
HPALF                40.28
HPALM                39.75
HP                   34.66
AMALF                31.46
AMALM                31.21
FRELCH               26.41
REDLCH               26.41
AM                   24.20
TOTFRL               23.59
ASALF                17.97
ASALM                17.66
AS                   13.73
BLALF                12.20
BLALM                11.52
FTE                   9.43
TRALF    

this(^) is important becuase you need to see how much missing data there is bc if its more than like 50% of missing data you will be getting rid of too much to get rid of it. you must find a new way to handle the missing values.

I see that the many of spaces where thera re are missing values is due to the grades of those students are not repsenented in the school grade limitattions. I will not be keeping those columns becuase I will be seeing the grade levels through the "school level" column. 


how many missinf values are there?
are these concentrated in specfic rows or coulmns? How does it affect the analysis





In [99]:
df['STUTERATIO'].describe()

count    99506.000000
mean        14.370399
std         60.800319
min         -2.000000
25%         11.000000
50%         14.030000
75%         17.100000
max      17900.000000
Name: STUTERATIO, dtype: float64

## Cleaning the Data

getting rid of the columns I dont need

In [100]:
columns_needed = [
    "NCESSCH", "LEAID", "ST_LEAID", "LEA_NAME", "SCH_NAME", "STABR", "LATCOD", "LONCOD","NMCNTY", "LCITY", "LSTATE", "LZIP",  # ID & Location
    
    "ULOCALE", "STATUS", "TOTFRL", "TOTAL", "STUTERATIO", "CHARTER_TEXT", "VIRTUAL", "FTE",  # School characteristics
   
    "SCHOOL_TYPE_TEXT", "SCHOOL_LEVEL", "CHARTER_TEXT", "VIRTUAL", "SY_STATUS_TEXT",  # School type 
    
    "AM", "AS", "BL", "HI", "WH", "HP", "TR"  # Racial categories
]

ps_data= df[columns_needed]

ps_data.head()

NCESSCH   LEAID ST_LEAID          LEA_NAME  \
0  10000500870  100005   AL-101  Albertville City   
1  10000500871  100005   AL-101  Albertville City   
2  10000500879  100005   AL-101  Albertville City   
3  10000500889  100005   AL-101  Albertville City   
4  10000501616  100005   AL-101  Albertville City   

                            SCH_NAME STABR   LATCOD     LONCOD  \
0          Albertville Middle School    AL  34.2602 -86.206200   
1            Albertville High School    AL  34.2622 -86.204900   
2    Albertville Intermediate School    AL  34.2733 -86.220100   
3      Albertville Elementary School    AL  34.2527 -86.221806   
4  Albertville Kindergarten and PreK    AL  34.2898 -86.193300   

            NMCNTY        LCITY LSTATE   LZIP           ULOCALE  STATUS  \
0  Marshall County  Albertville     AL  35950  32-Town: Distant       1   
1  Marshall County  Albertville     AL  35950  32-Town: Distant       1   
2  Marshall County  Albertville     AL  35950  32-Town: Distant       1   
3  Marshall County  Albertville     AL  35950  32-Town: Distant       1   
4  Marshall County  Albertville     AL  35951  32-Town: Distant       1   

   TOTFRL   TOTAL  STUTERATIO CHARTER_TEXT      VIRTUAL   FTE  \
0   332.0   908.0       21.62           No  Not Virtual  42.0   
1   456.0  1606.0       19.59           No  Not Virtual  82.0   
2   330.0   891.0       21.73           No  Not Virtual  41.0   
3   397.0   908.0       18.53           No  Not Virtual  49.0   
4   181.0   556.0       18.53           No  Not Virtual  30.0   

  SCHOOL_TYPE_TEXT SCHOOL_LEVEL CHARTER_TEXT      VIRTUAL  \
0   Regular school       Middle           No  Not Virtual   
1   Regular school         High           No  Not Virtual   
2   Regular school       Middle           No  Not Virtual   
3   Regular school   Elementary           No  Not Virtual   
4   Regular school   Elementary           No  Not Virtual   

           SY_STATUS_TEXT   AM   AS    BL     HI     WH   HP    TR  
0  Currently operational   2.0  4.0  33.0  469.0  371.0  0.0  29.0  
1  Currently operational   1.0  6.0  70.0  785.0  706.0  0.0  38.0  
2  Currently operational   4.0  7.0  24.0  481.0  345.0  1.0  29.0  
3  Currently operational   4.0  4.0  48.0  497.0  340.0  0.0  15.0  
4  Currently operational   4.0  2.0  33.0  288.0  216.0  0.0  13.0

In [101]:
# Im trying to see what values are not 1-- meaning school that are not operational
# I will get rid of any vaues that are not "1"

ps_data[ps_data['STATUS'] != 1].head(20)

NCESSCH   LEAID ST_LEAID                       LEA_NAME  \
154  10018802514  100188   AL-121                 Chickasaw City   
172  10019502498  100195   AL-186                 Pike Road City   
182  10020302516  100203   AL-802                   LEAD Academy   
183  10020402476  100204   AL-805                     i3 Academy   
184  10020402510  100204   AL-805                     i3 Academy   
185  10020702502  100207   AL-807                   LIFE Academy   
186  10020802503  100208   AL-808    Breakthrough Charter School   
187  10020902504  100209   AL-809          Ivy Classical Academy   
201  10021102505  100211   AL-810  Magic City Acceptance Academy   
261  10027002492  100270   AL-002                 Baldwin County   
396  10060002515  100600   AL-009                Chambers County   
523  10108002497  101080   AL-126                 Daleville City   
592  10123002512  101230   AL-130                    Dothan City   
593  10123002513  101230   AL-130                    Dothan City   
612  10129002494  101290   AL-026                  Elmore County   
690  10156002518  101560   AL-143                Fort Payne City   
778  10177002506  101770   AL-035                 Houston County   
779  10177002507  101770   AL-035                 Houston County   
780  10177002508  101770   AL-035                 Houston County   
781  10177002511  101770   AL-035                 Houston County   

                                              SCH_NAME STABR     LATCOD  \
154                Alabama Destinations Career Academy    AL  30.757103   
172                           Pike Road Jr High School    AL  32.355003   
182                            LEAD Academy Building B    AL  32.341405   
183                                         i3 Academy    AL  33.537508   
184                                 i3 Academy Phase 2    AL  33.564945   
185  LIFE Academy at Historic St Jude Educational I...    AL  32.352192   
186                        Breakthrough Charter School    AL  32.607571   
187                              Ivy Classical Academy    AL  32.459712   
201                      Magic City Acceptance Academy    AL  33.516759   
261                        Belforest Elementary School    AL  30.603160   
396                            Inspire Virtual Academy    AL  33.017341   
523                            Daleville Middle School    AL  31.315377   
592                         Dothan City Virtual School    AL  31.194437   
593                           Carver 9th Grade Academy    AL  31.194437   
612                                EDGE Virtual School    AL  32.546314   
690                   Little Ridge Intermediate School    AL  34.487696   
778                     Houston County Virtual Academy    AL  31.225600   
779                        Wicksburg Elementary School    AL  31.225600   
780                       Cottonwood Elementary School    AL  31.056759   
781                              Ashford Middle School    AL  31.183743   

        LONCOD             NMCNTY        LCITY LSTATE   LZIP  \
154 -88.078474      Mobile County    Chickasaw     AL  36611   
172 -86.102682  Montgomery County    Pike Road     AL  36064   
182 -86.223619  Montgomery County   Montgomery     AL  36116   
183 -86.763339   Jefferson County   Birmingham     AL  35212   
184 -86.726303   Jefferson County   Birmingham     AL  35206   
185 -86.328259  Montgomery County   Montgomery     AL  36108   
186 -87.335131       Perry County       Marion     AL  36756   
187 -86.470155     Autauga County   Prattville     AL  36067   
201 -86.784309   Jefferson County   Birmingham     AL  35233   
261 -87.829107     Baldwin County       Daphne     AL  36526   
396 -85.343690    Chambers County  Five Points     AL  36855   
523 -85.707641        Dale County     Dalevile     AL  36322   
592 -85.353124     Houston County       Dothan     AL  36301   
593 -85.353124     Houston County       Dothan     AL  36301   
612 -86.215789      Elmore County      Weump

In [102]:
# get rid of rows that contain -1. It was indicated that -1 in this dataset means the data is missing 


#seeing the -1 values
ps_data_missing = ps_data[(ps_data == -1).any(axis=1)]

ps_data_missing.head()






NCESSCH   LEAID ST_LEAID         LEA_NAME  \
8   10000600876  100006   AL-048  Marshall County   
14  10000600986  100006   AL-048  Marshall County   
15  10000600987  100006   AL-048  Marshall County   
37  10000702359  100007   AL-158      Hoover City   
38  10000702487  100007   AL-158      Hoover City   

                               SCH_NAME STABR     LATCOD     LONCOD  \
8                     Claysville School    AL  34.406900 -86.270400   
14          Marshall Alternative School    AL  34.406900 -86.270400   
15            Marshall Technical School    AL  34.305400 -86.285900   
37                    Crossroads School    AL  33.337604 -86.840932   
38  Riverchase Career Connection Center    AL  33.346680 -86.809690   

              NMCNTY         LCITY LSTATE   LZIP           ULOCALE  STATUS  \
8    Marshall County  Guntersville     AL  35976  41-Rural: Fringe       1   
14   Marshall County  Guntersville     AL  35976  41-Rural: Fringe       1   
15   Marshall County  Guntersville     AL  35976  41-Rural: Fringe       1   
37  Jefferson County        Hoover     AL  35244    13-City: Small       1   
38     Shelby County        Hoover     AL  35244    13-City: Small       1   

    TOTFRL  TOTAL  STUTERATIO CHARTER_TEXT      VIRTUAL  FTE  \
8      NaN    5.0        -1.0           No  Not Virtual  NaN   
14     NaN    NaN        -1.0           No  Not Virtual  NaN   
15     NaN    NaN        -1.0           No  Not Virtual  NaN   
37     NaN    NaN        -1.0           No  Not Virtual  NaN   
38     NaN    NaN        -1.0           No  Not Virtual  NaN   

                SCHOOL_TYPE_TEXT     SCHOOL_LEVEL CHARTER_TEXT      VIRTUAL  \
8                 Regular school  Prekindergarten           No  Not Virtual   
14  Alternative Education School            Other           No  Not Virtual   
15   Career and Technical School             High           No  Not Virtual   
37  Alternative Education School             High           No  Not Virtual   
38   Career and Technical School             High           No  Not Virtual   

            SY_STATUS_TEXT   AM   AS   BL   HI   WH   HP   TR  
8   Currently operational   0.0  0.0  0.0  2.0  3.0  0.0  0.0  
14  Currently operational   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
15  Currently operational   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
37  Currently operational   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
38  Currently operational   NaN  NaN  NaN  NaN  NaN  NaN  NaN

In [ ]:
#this code below will update my dataset to not include the rows with -1 values
ps_data_cleaned= ps_data[~(ps_data == -1).any(axis=1)]

ps_data_cleaned.head(50)

# I printed so many rows so I can see any patterns 

NCESSCH   LEAID ST_LEAID          LEA_NAME  \
0   10000500870  100005   AL-101  Albertville City   
1   10000500871  100005   AL-101  Albertville City   
2   10000500879  100005   AL-101  Albertville City   
3   10000500889  100005   AL-101  Albertville City   
4   10000501616  100005   AL-101  Albertville City   
5   10000502150  100005   AL-101  Albertville City   
6   10000600193  100006   AL-048   Marshall County   
7   10000600872  100006   AL-048   Marshall County   
9   10000600877  100006   AL-048   Marshall County   
10  10000600878  100006   AL-048   Marshall County   
11  10000600880  100006   AL-048   Marshall County   
12  10000600883  100006   AL-048   Marshall County   
13  10000600887  100006   AL-048   Marshall County   
16  10000601413  100006   AL-048   Marshall County   
17  10000601585  100006   AL-048   Marshall County   
18  10000601685  100006   AL-048   Marshall County   
19  10000601812  100006   AL-048   Marshall County   
20  10000602209  100006   AL-048   Marshall County   
21  10000700091  100007   AL-158       Hoover City   
22  10000700248  100007   AL-158       Hoover City   
23  10000700251  100007   AL-158       Hoover City   
24  10000700337  100007   AL-158       Hoover City   
25  10000700342  100007   AL-158       Hoover City   
26  10000701422  100007   AL-158       Hoover City   
27  10000701456  100007   AL-158       Hoover City   
28  10000701483  100007   AL-158       Hoover City   
29  10000701738  100007   AL-158       Hoover City   
30  10000701739  100007   AL-158       Hoover City   
31  10000701740  100007   AL-158       Hoover City   
32  10000701741  100007   AL-158       Hoover City   
33  10000701742  100007   AL-158       Hoover City   
34  10000701743  100007   AL-158       Hoover City   
35  10000702091  100007   AL-158       Hoover City   
36  10000702186  100007   AL-158       Hoover City   
39  10000800303  100008   AL-169      Madison City   
40  10000800495  100008   AL-169      Madison City   
41  10000800831  100008   AL-169      Madison City   
42  10000800839  100008   AL-169      Madison City   
43  10000800851  100008   AL-169      Madison City   
44  10000801423  100008   AL-169      Madison City   
45  10000801485  100008   AL-169      Madison City   
46  10000801798  100008   AL-169      Madison City   
47  10000802097  100008   AL-169      Madison City   
48  10000802133  100008   AL-169      Madison City   
49  10000802198  100008   AL-169      Madison City   
55  10001102094  100011   AL-167        Leeds City   
56  10001102095  100011   AL-167        Leeds City   
57  10001102096  100011   AL-167        Leeds City   
58  10001102430  100011   AL-167        Leeds City   
59  10001200123  100012   AL-115         Boaz City   

                                SCH_NAME STABR     LATCOD     LONCOD  \
0              Albertville Middle School    AL  34.260200 -86.206200   
1                Albertville High School    AL  34.262200 -86.204900   
2        Albertville Intermediate School    AL  34.273300 -86.220100   
3          Albertville Elementary School    AL  34.252700 -86.221806   
4      Albertville Kindergarten and PreK    AL  34.289800 -86.193300   
5             Albertville Primary School    AL  34.253300 -86.221800   
6           Kate Duncan Smith DAR Middle    AL  34.533700 -86.254100   
7                     Asbury High School    AL  34.362500 -86.142000   
9              Douglas Elementary School    AL  34.176234 -86.321259   
10                   Douglas High School    AL  34.178658 -86.319462   
11   Brindlee Mountain Elementary School    AL  34.344500 -86.442100   
12          Kate D Smith DAR High School    AL  34.534948 -86.253261   
13      Brindlee Mountain Primary School    AL  34.400054 -86.446830   
16               Robert D Sloman Primary    AL  34.176774 -86.322507   
17         Brindlee Mountain High School    AL  34.376400 -86.421900   
18    Kate D Smith DAR Elementary School    AL  34.534300 -86.254200   
19                 Douglas Mi

In [104]:
# Renaming my columns so I know what they are:
ps_data = ps_data_cleaned.rename(columns={
    "NCESSCH": "school_id",
    "LEAID": "district_id",
    "LEA_NAME": "district_name",
    "SCH_NAME": "school_name",
    "STABR": "state",
    
    "ULOCALE": "urban_locale",
    "SY_STATUS_TEXT": "school_status",
    "TOTFRL": "free_reduced_lunch",
    "TOTAL": "total_students",
    "STUTERATIO": "student_teacher_ratio",
    "CHARTER_TEXT": "charter_status",
    "VIRTUAL": "virtual_status",
    
    "SCHOOL_TYPE_TEXT": "school_type",
    "SCHOOL_LEVEL": "school_level",
    "NMCNTY": "county_name",
    "LCITY": "city",
    "LSTATE": "state_full",
    "LZIP": "zip_code",
    
    "AM": "american_indian",
    "AS": "asian",
    "BL": "black",
    "HI": "hispanic",
    "WH": "white",
    "HP": "pacific_islander",
    "TR": "two_or_more_races"
})

In [105]:
# I need to get rid of rows that dont have "NA" values int the TOTAL

# seeing how many missing values are in the TOTFRL column

num_missing_total= ps_data["total_students"].isna().sum()

print(f" total number of nan values in the total student column: {num_missing_total}")

pct_missing_total= (ps_data["total_students"].isna().sum() / len(ps_data)) * 100

print(f" percent of nan values in the total_student column: {pct_missing_total: .2f}%")

 total number of nan values in the total student column: 1042
 percent of nan values in the total_student column:  1.13%


#### I will get rid of those NaN rows in "total_students" since there arent that many rows with those missue values and I need an integer in the total_students column for my future analysis

In [106]:
#getting rid of the NaN values in "total_students" column
ps_data= ps_data.dropna(subset= ["total_students"])




In [107]:
#I need to get rid of rows where the total stidents == 0. I dont need that information

zero_total= (ps_data["total_students"] == 0).sum()

print(zero_total)

1497


In [108]:
ps_data=ps_data[ps_data["total_students"] != 0]

ps_data["total_students"].describe()


count    89601.000000
mean       518.561757
std        471.374405
min          1.000000
25%        255.000000
50%        421.000000
75%        636.000000
max      21049.000000
Name: total_students, dtype: float64

In [109]:
high_total= ps_data[ps_data["total_students"]== ps_data["total_students"].max()]

high_total

# I may need to get rid of fully virtual schools. The total student value for this school(the max) is too high and may skew my data

school_id  district_id   ST_LEAID         district_name  \
67048  390018004704      3900180  OH-142950  Ohio Virtual Academy   

                school_name state     LATCOD     LONCOD   county_name    city  \
67048  Ohio Virtual Academy    OH  41.587027 -83.684248  Lucas County  Maumee   

      state_full  zip_code      urban_locale  STATUS  free_reduced_lunch  \
67048         OH     43537  21-Suburb: Large       1                 NaN   

       total_students  student_teacher_ratio charter_status virtual_status  \
67048         21049.0                  50.97            Yes   Full Virtual   

         FTE     school_type school_level charter_status virtual_status  \
67048  413.0  Regular school        Other            Yes   Full Virtual   

                school_status  american_indian  asian   black  hispanic  \
67048  Currently operational              63.0  224.0  3160.0    1603.0   

         white  pacific_islander  two_or_more_races  
67048  14297.0              15.0             1687.0

In [110]:
# -9 means that the data do not meet NCES data quality standards.

ps_data.describe()

school_id   district_id        LATCOD        LONCOD      zip_code  \
count  8.960100e+04  8.960100e+04  89601.000000  89601.000000  89601.000000   
mean   2.938175e+11  2.938175e+06     37.731886    -92.783465  52836.831832   
std    1.656855e+11  1.656855e+06      5.766947     17.302496  29721.891712   
min    1.000050e+10  1.000050e+05     13.272449   -176.640331    601.000000   
25%    1.301500e+11  1.301500e+06     33.921600   -101.187600  28787.000000   
50%    3.003870e+11  3.003870e+06     38.514696    -89.474800  51106.000000   
75%    4.218990e+11  4.218990e+06     41.573040    -80.430000  78521.000000   
max    7.800030e+11  7.800030e+06     71.300337    144.917553  99950.000000   

             STATUS  free_reduced_lunch  total_students  \
count  89601.000000        75068.000000    89601.000000   
mean       1.018928          283.358209      518.561757   
std        0.222353          294.584577      471.374405   
min        1.000000            3.000000        1.000000   
25%        1.000000           96.000000      255.000000   
50%        1.000000          209.000000      421.000000   
75%        1.000000          377.000000      636.000000   
max        8.000000        10595.000000    21049.000000   

       student_teacher_ratio           FTE  american_indian         asian  \
count           89132.000000  89601.000000     70119.000000  79492.000000   
mean               16.160761     33.482417         6.764857     31.790394   
std                64.001335     25.839377        29.768138     90.288846   
min                 0.000000      0.000000         0.000000      0.000000   
25%                12.130000     18.000000         0.000000      1.000000   
50%                14.570000     28.180000         1.000000      6.000000   
75%                17.560000     41.709999         4.000000     24.000000   
max             17900.000000    683.270020      1308.000000   3517.000000   

              black      hispanic         white  pacific_islander  \
count  83749.000000  88351.000000  88247.000000      60287.000000   
mean      83.659709    150.075902    236.928632          3.080133   
std      152.648125    245.203241    274.225781         21.967201   
min        0.000000      0.000000      0.000000          0.000000   
25%        4.000000     15.000000     47.000000          0.000000   
50%       21.000000     56.000000    172.000000          1.000000   
75%       97.000000    187.000000    331.000000          2.000000   
max     5360.000000   4461.000000  14297.000000       1301.000000   

       two_or_more_races  
count       85596.000000  
mean           24.423314  
std            35.378035  
min             0.000000  
25%             5.000000  
50%            15.000000  
75%            33.000000  
max          3538.000000

In [111]:
# adding a column that shows the %

pct_columns = ["black",
    "white",
    "hispanic",
    "american_indian",
    "pacific_islander",
    "two_or_more_races",
    "asian"]

for col in pct_columns:
    if col in ps_data.columns:
        ps_data[col + "_pct"] = (ps_data[col]/ ps_data ["total_students"])* 100

ps_data.head()




school_id  district_id ST_LEAID     district_name  \
0  10000500870       100005   AL-101  Albertville City   
1  10000500871       100005   AL-101  Albertville City   
2  10000500879       100005   AL-101  Albertville City   
3  10000500889       100005   AL-101  Albertville City   
4  10000501616       100005   AL-101  Albertville City   

                         school_name state   LATCOD     LONCOD  \
0          Albertville Middle School    AL  34.2602 -86.206200   
1            Albertville High School    AL  34.2622 -86.204900   
2    Albertville Intermediate School    AL  34.2733 -86.220100   
3      Albertville Elementary School    AL  34.2527 -86.221806   
4  Albertville Kindergarten and PreK    AL  34.2898 -86.193300   

       county_name         city state_full  zip_code      urban_locale  \
0  Marshall County  Albertville         AL     35950  32-Town: Distant   
1  Marshall County  Albertville         AL     35950  32-Town: Distant   
2  Marshall County  Albertville         AL     35950  32-Town: Distant   
3  Marshall County  Albertville         AL     35950  32-Town: Distant   
4  Marshall County  Albertville         AL     35951  32-Town: Distant   

   STATUS  free_reduced_lunch  total_students  student_teacher_ratio  \
0       1               332.0           908.0                  21.62   
1       1               456.0          1606.0                  19.59   
2       1               330.0           891.0                  21.73   
3       1               397.0           908.0                  18.53   
4       1               181.0           556.0                  18.53   

  charter_status virtual_status   FTE     school_type school_level  \
0             No    Not Virtual  42.0  Regular school       Middle   
1             No    Not Virtual  82.0  Regular school         High   
2             No    Not Virtual  41.0  Regular school       Middle   
3             No    Not Virtual  49.0  Regular school   Elementary   
4             No    Not Virtual  30.0  Regular school   Elementary   

  charter_status virtual_status           school_status  american_indian  \
0             No    Not Virtual  Currently operational               2.0   
1             No    Not Virtual  Currently operational               1.0   
2             No    Not Virtual  Currently operational               4.0   
3             No    Not Virtual  Currently operational               4.0   
4             No    Not Virtual  Currently operational               4.0   

   asian  black  hispanic  white  pacific_islander  two_or_more_races  \
0    4.0   33.0     469.0  371.0               0.0               29.0   
1    6.0   70.0     785.0  706.0               0.0               38.0   
2    7.0   24.0     481.0  345.0               1.0               29.0   
3    4.0   48.0     497.0  340.0               0.0               15.0   
4    2.0   33.0     288.0  216.0               0.0               13.0   

   black_pct  white_pct  hispanic_pct  american_indian_pct  \
0   3.634361  40.859031     51.651982             0.220264   
1   4.358655  43.960149     48.879203             0.062267   
2   2.693603  38.720539     53.984287             0.448934   
3   5.286344  37.444934     54.735683             0.440529   
4   5.935252  38.848921     51.798561             0.719424   

   pacific_islander_pct  two_or_more_races_pct  asian_pct  
0              0.000000               3.193833   0.440529  
1              0.000000               2.366127   0.373599  
2              0.112233               3.254770   0.785634  
3              0.000000               1.651982   0.440529  
4              0.000000               2.338129   0.359712

I am seeing there is missing data in all colmumns when there is a -1 in free_reduced_lunch-- get rid of those rows

In [118]:
#seeing the placements of all the schools 
ps_data["urban_locale"].value_counts()

urban_locale
21-Suburb: Large       23207
11-City: Large         13429
41-Rural: Fringe       10571
42-Rural: Distant       9165
43-Rural: Remote        5908
13-City: Small          5716
12-City: Mid-size       5317
32-Town: Distant        5268
33-Town: Remote         3527
22-Suburb: Mid-size     3072
31-Town: Fringe         2645
23-Suburb: Small        1776
Name: count, dtype: int64